### DISCUSSION TITLE
Dump to dict with Decimal instead of float

### DISCUSSION THREAD

**Discussion by mdavis-xyz** [pydantic/pydantic#4701] [Pydantic V2]
One of the main inputs and outputs of my scripts that use pydantic is AWS' DynamoDB no-sql database.

The `boto3` SDK [only handles `Decimal` (and `int`), not `float`](https://github.com/boto/boto3/issues/369#issuecomment-793048012). When I read from dynamodb it gives me `Decimal`, and pydantic can coerce that into `float`, which is fantastic.
But if I want to write a pydantic model to dynamodb, I need to convert all `float`s to `Decimal`. (I also need to convert datetimes to strings). This is not straightforward with pydantic v1.

I currently subclass `BaseModel`, and then every other model subclasses that.

```
class BaseModel(pydantic.BaseModel):
    # decimal=True means convert all floats to Decimal
    # primitives=True means convert enums, uuids, datetimes etc all to str
    def dict(self, decimal=False, primitives=False, exclude_unset=True, exclude_none=True, **kwargs):
        if primitives:
            # just dump to json, then parse back
            # a bit slow, but fairly reliable,
            # since we've already specified json encoders
            d = json.loads(self.json(exclude_unset=exclude_unset, exclude_none=exclude_none, **kwargs))
        else:
            d = super().dict(exclude_unset=exclude_unset, exclude_none=exclude_none, **kwargs)
        if decimal:
            d = float_to_decimal(d)
            
        return d

    class Config:
        use_enum_values=True # not sure if this works recursively
        validate_assignment = True 
        json_encoders = {
            dt.datetime: encode_datetime,
            dt.time: encode_time,
            dt.date: encode_date
            # TODO: add UUID here?
        }
```

It would be great if I could dump my model into a dict that contains Decimals not floats. (And strings not datetimes.)

**Comment by samuelcolvin** [comment]
Not easily possible in v1, but should be easy in V2.

**Reply by adriangb** [reply]
Here's how  I would do it if I'm understanding your use case correctly:

```python
from decimal import Decimal
from typing import Annotated

from pydantic import TypeAdapter, PlainSerializer

FloatAsDecimal = Annotated[float, PlainSerializer(lambda x: Decimal(x), return_type=Decimal)]

ta = TypeAdapter(FloatAsDecimal)

v = ta.dump_python(1.5)
assert type(v) is Decimal
assert v == Decimal(1.5)
```

**Reply by orfisko** [reply]
Can this be elaborated upon? I need a solution that works in models with deeply nested floats, together with all sorts of other data.

**Comment by bravesoul349** [comment]
I was facing same issue with boto3 `Float types are not supported. Use Decimal types instead.` and landed directly here therefore, leaving my solution for others.

```
class Home(BaseModel):
  long: Decimal
  lat: Decimal

home = Home(long=74.0059, lat=75.2668)

d = json.loads(home.json(), parse_float=Decimal)
```
